### PR TITLE
`Development`: Untangle quiz dependencies and improve code formatting

### PR DESCRIPTION
**Pull request by julian-christl**
#### General

- [X] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [X] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [X] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [X] I documented the Java code using JavaDoc style.

### Motivation and Context

We have to set a few more relations to lazy to break a loop that gets introduced by the Hibernate 6 upgrade.

### Description

- Updated fetch type for `DropLocation`, `ShortAnswerSolution`, `ShortAnswerSpot` 
- Improved some styling and comments
- Removed some entity service dependencies by setting FilePathService methods to `static`

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Create a Quiz Exercise with a Drag And Drop question
2. Test out preview and sample solution as the instructor
3. Participate in the quiz
4. Make sure the result is properly displayed

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **Refactor**
	- Improved readability of annotation formatting across various domain entities.
	- Updated method calls to utilize static methods from `FilePathService` for file path handling.
	- Added comments to clarify bidirectional and unidirectional relationships in quiz-related entities.
	- Changed fetch strategies to `FetchType.LAZY` in certain `@ManyToOne` annotations to optimize loading behavior.
	- Reordered annotations and reformatted code blocks for consistency and clarity.

- **New Features**
	- Introduced new repository methods for eager loading of quiz entities.

- **Bug Fixes**
	- Fixed a variable name typo in `ExerciseDeletionService`.

- **Documentation**
	- Added clarifying comments on Hibernate behavior and necessity of mappings in quiz entities.

- **Tests**
	- Refactored submission creation logic in integration tests for better maintainability.
	- Adapted test cases to reflect changes in `FilePathService` method calls.

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java [hunk 1]
```diff
@@ -1296,6 +1284,22 @@ else if (exercise instanceof QuizExercise) {
...
+        }
+        else if (exercise instanceof QuizExercise) {
+            return new QuizSubmission();
+        }
+        return null;
+    }
+
     @Test
     @WithMockUser(username = TEST_PREFIX + "instructor1", roles = "INSTRUCTOR")
     void testSubmitStudentExam_early() throws Exception {
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java:1305]: It gets used in line 1259

File: src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java [hunk 2]
```diff
@@ -1296,6 +1284,22 @@
...
+            return new TextSubmission();
+        }
+        else if (exercise instanceof ModelingExercise) {
+            return new ModelingSubmission();
+        }
+        else if (exercise instanceof QuizExercise) {
+            return new QuizSubmission();
+        }
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/java/de/tum/in/www1/artemis/exam/StudentExamIntegrationTest.java:1301]: It gets used in line 1259

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/ShortAnswerQuestion.java [hunk 3]
```diff
@@ -22,23 +22,29 @@
 @JsonInclude(JsonInclude.Include.NON_EMPTY)
 public class ShortAnswerQuestion extends QuizQuestion {
 
+    // TODO: making this a bidirectional relation leads to weired Hibernate behavior with missing data when loading quiz questions, we should investigate this again in the future
+    // after 6.x upgrade
     @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER, orphanRemoval = true)
-    @OrderColumn
     @JoinColumn(name = "question_id")
+    @OrderColumn
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/domain/quiz/ShortAnswerQuestion.java:29]: Fixed all occurrences in d2599fe1988a43969024a30a15b764be0be7d551

**Review by konrad2002 — APPROVED** [review]
manually tested on ts1, works as expected

**Review by LenaKahle — APPROVED** [review]
Approve, tested on legacy ts11.

File: src/main/java/de/tum/in/www1/artemis/domain/iris/session/IrisSession.java [hunk 4]
```diff
@@ -25,9 +25,11 @@
 @Inheritance(strategy = InheritanceType.SINGLE_TABLE)
 @DiscriminatorColumn(name = "discriminator", discriminatorType = DiscriminatorType.STRING)
 @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
+// @formatter:on
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
 @JsonSubTypes({ @JsonSubTypes.Type(value = IrisChatSession.class, name = "chat"), @JsonSubTypes.Type(value = IrisHestiaSession.class, name = "hestia"),
         @JsonSubTypes.Type(value = IrisCodeEditorSession.class, name = "codeEditor") })
+// @formatter:off
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/domain/iris/session/IrisSession.java:32]: I think you wanted it the other way around here.

File: src/main/java/de/tum/in/www1/artemis/repository/QuizQuestionRepository.java [hunk 5]
```diff
@@ -20,13 +21,20 @@ public interface QuizQuestionRepository extends JpaRepository<QuizQuestion, Long
...
-                LEFT JOIN  exercise.quizQuestions question
+                LEFT JOIN exercise.quizQuestions question
             WHERE exercise.id = :exerciseId
             """)
     Set<QuizQuestion> getQuizQuestionsByExerciseId(@Param("exerciseId") long exerciseId);
 
+    @Query("""
+            SELECT question
+                FROM DragAndDropQuestion question
+                WHERE question.id = :questionId
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/repository/QuizQuestionRepository.java:32]:
```suggestion
            SELECT question
            FROM DragAndDropQuestion question
            WHERE question.id = :questionId
```

File: src/main/java/de/tum/in/www1/artemis/service/AttachmentUnitService.java [hunk 6]
```diff
@@ -154,7 +154,7 @@ private void handleFile(MultipartFile file, Attachment attachment, boolean keepF
      */
     private void evictCache(MultipartFile file, AttachmentUnit attachmentUnit) {
         if (file != null && !file.isEmpty()) {
-            this.fileService.evictCacheForPath(filePathService.actualPathForPublicPathOrThrow(URI.create(attachmentUnit.getAttachment().getLink())));
+            this.fileService.evictCacheForPath(FilePathService.actualPathForPublicPathOrThrow(URI.create(attachmentUnit.getAttachment().getLink())));
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/AttachmentUnitService.java:157]: Can you now remove the `filePathService` attribute from this file?

File: src/main/java/de/tum/in/www1/artemis/service/LectureImportService.java [hunk 7]
```diff
@@ -162,7 +162,7 @@ private Attachment cloneAttachment(final Attachment importedAttachment) {
         attachment.setVersion(importedAttachment.getVersion());
         attachment.setAttachmentType(importedAttachment.getAttachmentType());
 
-        Path oldPath = filePathService.actualPathForPublicPathOrThrow(URI.create(importedAttachment.getLink()));
+        Path oldPath = FilePathService.actualPathForPublicPathOrThrow(URI.create(importedAttachment.getLink()));
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/LectureImportService.java:165]: There is another call of a static method in FilePathService below. After changing this you could also remove the filePathService attribute in this class.

File: src/main/java/de/tum/in/www1/artemis/service/QuizExerciseService.java [hunk 8]
```diff
@@ -323,7 +323,7 @@ public void validateQuizExerciseFiles(QuizExercise quizExercise, @Nonnull List<M
         Set<String> exerciseFileNames = getAllPathsFromDragAndDropQuestionsOfExercise(quizExercise);
         Set<String> newFileNames = isCreate ? exerciseFileNames : exerciseFileNames.stream().filter(fileNameOrUri -> {
             try {
-                return !Files.exists(filePathService.actualPathForPublicPathOrThrow(URI.create(fileNameOrUri)));
+                return !Files.exists(FilePathService.actualPathForPublicPathOrThrow(URI.create(fileNameOrUri)));
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/QuizExerciseService.java:326]: same here. (I won't comment on all affected classes now, please double-check that)

**Review by nata-81 — APPROVED** [review]
Tested on ts2 works as expected.

**Review by AntonGeTUM — APPROVED** [review]
Manual test on ts11, works as expected

**Review by Strohgelaender — DISMISSED** [review]
code lgtm

File: src/main/java/de/tum/in/www1/artemis/domain/submissionpolicy/SubmissionPolicy.java [hunk 9]
```diff
@@ -29,8 +29,10 @@
 @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
 @JsonInclude(JsonInclude.Include.NON_EMPTY)
+// @formatter:on
 @JsonSubTypes({ @JsonSubTypes.Type(value = LockRepositoryPolicy.class, name = "lock_repository"),
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/domain/submissionpolicy/SubmissionPolicy.java:33]:
```suggestion
@JsonSubTypes({
    @JsonSubTypes.Type(value = LockRepositoryPolicy.class, name = "lock_repository"),
```

File: src/main/java/de/tum/in/www1/artemis/domain/submissionpolicy/SubmissionPolicy.java [hunk 10]
```diff
@@ -29,8 +29,10 @@
...
         @JsonSubTypes.Type(value = SubmissionPenaltyPolicy.class, name = "submission_penalty") })
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/domain/submissionpolicy/SubmissionPolicy.java:34]:
```suggestion
        @JsonSubTypes.Type(value = SubmissionPenaltyPolicy.class, name = "submission_penalty")
})
```

**Review by MaximilianAnzinger — DISMISSED** [review]
Besides this minor inconsistency and my question the code looks great 👍

**Review by nata-81 — APPROVED** [review]
Re-tested on ts1, works as expected.

File: src/main/java/de/tum/in/www1/artemis/repository/QuizQuestionRepository.java [hunk 11]
```diff
@@ -20,13 +21,20 @@ public interface QuizQuestionRepository extends JpaRepository<QuizQuestion, Long
...
             WHERE exercise.id = :exerciseId
             """)
     Set<QuizQuestion> getQuizQuestionsByExerciseId(@Param("exerciseId") long exerciseId);
 
+    @Query("""
+            SELECT question
+            FROM DragAndDropQuestion question
+            WHERE question.id = :questionId
+            """)
+    Optional<DragAndDropQuestion> findDnDQuestionById(@Param("questionId") long questionId);
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/repository/QuizQuestionRepository.java:28]: No, we would have to create a dedicated repository for DragAndDrop questions as the findById method would override the method returning an Optional<QuizQuestion>. IMO it's unnecessary to create a wole new repo as a DragAndDropQuestion is a QuizQuestion.

File: src/main/java/de/tum/in/www1/artemis/service/export/DataExportQuizExerciseCreationService.java [hunk 12]
```diff
@@ -73,7 +73,7 @@ public void createQuizAnswersExport(QuizExercise quizExercise, StudentParticipat
      */
     private boolean createQuizAnswersExport(QuizExercise quizExercise, StudentParticipation participation, Path outputDir, boolean includeResults,
             Optional<List<String>> exportErrors) {
-        Set<QuizQuestion> quizQuestions = quizQuestionRepository.getQuizQuestionsByExerciseId(quizExercise.getId());
+        Set<QuizQuestion> quizQuestions = quizQuestionRepository.findByExercise_Id(quizExercise.getId());
         boolean errorOccurred = false;
         for (var submission : participation.getSubmissions()) {
             QuizSubmission quizSubmission = quizSubmissionRepository.findWithEagerSubmittedAnswersById(submission.getId());
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/service/export/DataExportQuizExerciseCreationService.java:79]: Out of scope

**Review by Strohgelaender — APPROVED** [review]

**Review by konrad2002 — APPROVED** [review]
re-approve, still working

**Review by AntonGeTUM — APPROVED** [review]
retested on ts11, approved

**Review by LenaKahle — APPROVED** [review]
Reapprove.

**Review by MaximilianAnzinger — APPROVED** [review]

**Pull request merged**